File: explores/linked-list/0206_reverse_linked_list.cpp
#include "ListNode.h"

class Solution {
   private:
    ListNode* reverseRecursion(ListNode* prev, ListNode* curr) {
        if (!curr) {
            return prev;
        }

        ListNode* next = curr->next;
        curr->next = prev;
        return reverseRecursion(curr, next);
    }

   public:
    ListNode* reverseList(ListNode* head) {
        // ListNode* prev = nullptr;
        // ListNode* curr = head;

        // while (curr) {
        //     ListNode* next = curr->next;
        //     curr->next = prev;
        //     prev = curr;
        //     curr = next;
        // }
        // return prev;
        return reverseRecursion(nullptr, head);
    }
};

File: explores/recursion-i/0206_reverse_linked_list.py
from typing import Optional

from ListNode import ListNode


class Solution:
    def rec(self, prev: Optional[ListNode], curr: Optional[ListNode]) -> Optional[ListNode]:
        if curr is None:
            return prev

        next = curr.next
        curr.next = prev
        prev = curr
        return self.rec(curr, next)

    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        return self.rec(None, head)
